File: LeetCode/cpp/990.cpp
class Solution {
public:
    bool equationsPossible(vector<string> &equations) {
        match = vector<int>(26, 0);
        for (int i = 0; i < 26; ++i)
            match[i] = i;
        for (auto &s:equations) {
            if (s[1] == '=')
                match[Match(s[0] - 'a')] = Match(s[3] - 'a');
        }
        for (auto &s:equations) {
            if (s[1] == '!')
                if (Match(match[s[0] - 'a']) == Match(match[s[3] - 'a']))
                    return false;
        }
        return true;
    }

private:
    vector<int> match;

    int Match(int i) {
        if (match[i] == i)
            return i;
        return Match(match[i]);
    }
};

File: LeetCode/python3/990.py
class Solution:
    match = []

    def Match(self, i):
        if i == self.match[i]:
            return i
        return self.Match(self.match[i])

    def equationsPossible(self, equations: 'List[str]') -> 'bool':
        self.match = []
        for i in range(26):
            self.match.append(i)
        for s in equations:
            if s[1] == '=':
                self.match[self.Match(ord(s[0]) - 97)] = self.Match(ord(s[3]) - 97)
        for s in equations:
            if s[1] == '!':
                if self.Match(ord(s[0]) - 97) == self.Match(ord(s[3]) - 97):
                    return False
        return True